pkg/cache/v3/resources.go

```diff
@@ -12,25 +16,34 @@ type Resources struct {
 }
 
 // IndexResourcesByName creates a map from the resource name to the resource.
-func IndexResourcesByName(items []types.ResourceWithTTL) map[string]types.ResourceWithTTL {
+func IndexResourcesByName(items []types.ResourceWithTTL) (map[string]types.ResourceWithTTL, error) {
```

Comment:
Do you still need to maintain hard compatibility for v0.x.x?

If it is a hard requirement to keep the interface we have a few options
1. Add the check inside NewSnapshot. 
2. We can also deprecate the existing functions and create internal versions of them with the new interface. 

For both these approaches the behaviour changes, but interface is kept. We can also guard behaviour behind an environment variable to allow users to switch back to the old behaviour. IMO this a worst option, since this behaviour change might not get caught before deploying to production. We could also keep the old behaviour by default, but the issue is it currently fails silently and the main benefit of this change is better observability.

Reply:
Do you still need to maintain hard compatibility for v0.x.x?

If it is a hard requirement to keep the interface we have a few options
1. Add the check inside NewSnapshot. 
2. We can also deprecate the existing functions and create internal versions of them with the new interface. 

For both these approaches the behaviour changes, but interface is kept. We can also guard behaviour behind an environment variable to allow users to switch back to the old behaviour. IMO this a worst option, since this behaviour change might not get caught before deploying to production. We could also keep the old behaviour by default, but the issue is it currently fails silently and the main benefit of this change is better observability.